src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -26,23 +26,65 @@
 package jdk.incubator.foreign;
 
 import jdk.internal.foreign.NativeMemorySegmentImpl;
+import jdk.internal.foreign.PlatformLayouts;
 import jdk.internal.foreign.Utils;
 import jdk.internal.foreign.abi.SharedUtils;
 
-import java.nio.ByteOrder;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodType;
 import java.nio.charset.Charset;
 import java.util.Objects;
 import java.util.function.Consumer;
 
+import static jdk.internal.foreign.PlatformLayouts.*;
+
 /**
- * A set of utilities for working with libraries using the C language/ABI
+ * This interface is a used as a super class for all ForeinLinker implementations that
+ * implement the system's C ABI and calling convention. As such, instances of this interface
+ * can be used to link foreign functions in native libraries that follow the JVM's target
+ * platform C application binary interface (ABI).
+ *
+ * <p>There are two components that go into linking a foreign function: a method type, and
+ * a function descriptor. The method type, consisting of a set of 'carrier' types constitutes
+ * the Java side of a call to a foreign function, while the function descriptor constitutes
+ * the native side of the call. Memory layout attributes are used in the function descriptor
+ * to attach ABI classification meta-data to memory layouts, which are required for linking.
+ * Clients of this API should use the prepared memory layout constants found in this interface
+ * to create their function descriptor, based on the built-in types provided by the C language.</p>
+ *
+ * <p>Implementations of this interface support the following primitive carrier types:
+ * {@code byte}, {@code short}, {@code char}, {@code int}, {@code long}, {@code float},
+ * and {@code double}, as well as {@link MemoryAddress} for passing pointers, and
+ * {@link MemorySegment} for passing structs and unions. Finally, the {@link VaList}
+ * carrier type can be used to match the native {@code va_list} type.</p>
+ *
+ * <p>The function descriptor used in linking contains a memory layout to match each carrier type.
+ * There are some restrictions on the carrier type and memory layout combinations that are allowed:</p>
+ * <ul>
+ *   <li>If a primitve type is used as a carrier type, the corresponding
+ *   memory layout must be a {@code ValueLayout}, and the bit size of the layout must match that of the carrier type
+ *   (see {@link Integer#SIZE} and similar fields in other primitive wrapper classes).</li>
+ *
+ *   <li>If the carrier type is {@code MemoryAddress} or {@code VaList}, then the corresponding memory layout must be a
```

Comment:
I would split this into two, one for pointers and one for VaList since the carrier and layout is different